### PR TITLE
Allow quota key with different auth scheme in HTTPHandler

### PR DESCRIPTION
**Pull request by kssenii**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Allow quota key with different auth scheme in HTTP requests.

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

---
### Modify your CI run:
**NOTE:** If your merge the PR with modified CI you **MUST KNOW** what you are doing
**NOTE:** Checked options will be applied if set before CI RunConfig/PrepareRunConfig step

#### Include tests (required builds will be added automatically):
- [ ]  Fast test
- [ ]  Integration Tests
- [ ]  Stateless tests
- [ ]  Stateful tests
- [ ]  Unit tests
- [ ]  Performance tests
- [ ]  All with ASAN
- [ ]  All with TSAN
- [ ]  All with Analyzer
- [ ]  Add your option here

#### Exclude tests:
- [ ]  Fast test
- [ ]  Integration Tests
- [ ]  Stateless tests
- [ ]  Stateful tests
- [ ]  Performance tests
- [ ]  All with ASAN
- [ ]  All with TSAN
- [ ]  All with MSAN
- [ ]  All with UBSAN
- [ ]  All with Coverage
- [ ]  All with Aarch64
- [ ]  Add your option here

#### Extra options:
- [ ]  do not test (only style check)
- [ ]  disable merge-commit (no merge from master before tests)
- [ ]  disable CI cache (job reuse)

#### Only specified batches in multi-batch jobs:
- [ ]  1
- [ ]  2
- [ ]  3
- [ ]  4

### CODE REVIEW

**Comment by robot-clickhouse** [comment]
*This is an automated comment for commit 4a68b2a877bc30322552031535ede4247e08385b with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/62842/4a68b2a877bc30322552031535ede4247e08385b/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

**Review by vitlibar — APPROVED** [review]

File: src/Server/HTTPHandler.cpp [hunk 1]
```diff
@@ -364,16 +364,26 @@ bool HTTPHandler::authenticateUser(
...
 
+    if (params.has("quota_key"))
+    {
+        if (!quota_key.empty())
+            throw Exception(ErrorCodes::AUTHENTICATION_FAILED,
+                            "Invalid authentication: it is not allowed "
+                            "to use quota key as HTTP header and quota key as parameter simultaneously");
+
+        quota_key = params.get("quota_key");
+    }
```
Comment (vitlibar) [inline src/Server/HTTPHandler.cpp:385]: I think it's better to move this quota-specific code later near the place `quota_key` is used.
Also it seems the error code here should be rather something like `BAD_ARGUMENTS`.

File: tests/queries/0_stateless/01399_http_request_headers.sh [hunk 2]
```diff
@@ -15,3 +15,4 @@ ${CLICKHOUSE_CURL} -sS "${CLICKHOUSE_URL}" -H 'X-ClickHouse-Database: system' -d
 ${CLICKHOUSE_CURL} -sS "${CLICKHOUSE_URL}" -H 'X-ClickHouse-Database: header_test' -d 'SHOW TABLES' | grep -o 'Code: 81'
 ${CLICKHOUSE_CURL} -sS "${CLICKHOUSE_URL}" -H 'X-ClickHouse-Format: JSONCompactEachRow' -d 'SELECT 1' | grep -o '\[1\]'
 ${CLICKHOUSE_CURL} -sS "${CLICKHOUSE_URL}" -H 'X-ClickHouse-Format: header_test' -d 'SELECT 1' | grep -o 'Code: 73'
+${CLICKHOUSE_CURL} -sS "${CLICKHOUSE_URL}&quota_key=pingpong" -H 'Authorization: Basic dHQ6cGFzc3dvcmQ=' -d 'SELECT 1' | grep -o 'Code: 516'
```
Comment (vitlibar) [inline tests/queries/0_stateless/01399_http_request_headers.sh:18]: I don't understand this test, it seems this `SELECT 1` should work normally now with your changes.

Comment (kssenii) [reply tests/queries/0_stateless/01399_http_request_headers.sh]: fixed

**Comment by kssenii** [comment]
> Stateless tests

03000_traverse_shadow_system_data_paths
00857_global_joinsavel_table_alias

-- not related to changes

> AST fuzzer (msan)

https://github.com/ClickHouse/ClickHouse/issues/61722

**Comment by clickgapai** [comment]
Hi @kssenii @vitlibar — while reviewing this PR I found the following:

- **Coverage gap:** New `BAD_ARGUMENTS` error path for simultaneous `X-ClickHouse-Quota` header and `quota_key` URL parameter is untested — https://github.com/ClickHouse/ClickHouse/pull/104092

Could you review the test PR and add the `can be tested` label if it looks good?
Happy to discuss — close anything that's wrong or already addressed.

**Pull request merged**